### DISCUSSION TITLE
Does https://github.com/libuv/libuv/pull/2231 fix "use after free" in all possible path?

### DISCUSSION THREAD

**Discussion by xiaoxiang781216** [libuv/libuv#3627] [General]
Here is an example:

1. uv__async_close just pass uv__async_spin in the looper thread
2. Work thread call uv_async_send for the same handle at the same time
3. Looper thread call close_cb which may release the handle before uv_async_send return

So, the work thread may touch the freed memory. @bnoordhuis how to handle this race condition?

**Comment by vtjnash** [comment]
So don’t do that? The user should not be calling close if the handle is still in use elsewhere in the program

**Reply by xiaoxiang781216** [reply]
PR #2231 can't fix the problem it want, so should we revert the change and let's caller handle the issue by self? I am asking this question because we found PR #2231 introduce the busy deadlock in the real time operating system(https://github.com/apache/incubator-nuttx) we are using like this:

1. Low priority thread call uv_async_send and just change pending from 0 to 1 and pass uv__async_send
2. Scheduler switch to loop thread and enter uv__async_io before low priority thread change pending from 1 to 2
3. Loop thread is busy in uv__async_spin to wait pending doesn't equals 1, but it's never happen

Even sched_yield added in PR #2772 can't totally resolve this issue, because:

1. sched_yield never give the time slice to the low priority thread from:
    https://man7.org/linux/man-pages/man2/sched_yield.2.html 
2. If looper thread use SCHED_FIFO policy(our case), scheduler never switch to the low priority thread from:
    https://man7.org/linux/man-pages/man2/sched_yield.2.html

So, it isn't an issue specific to NuttX, but to all POSIX compliant OS(FreeBSD, Linux...).

**Comment by xiaoxiang781216** [comment]
BTW, https://github.com/libuv/libuv/pull/3015 report the short busy loop.
So, I am wondering if #2231 can't fix the problem, why not revert them to avoid the performance lose or even the dead lock? @vtjnash @bnoordhuis .

**Reply by bnoordhuis** [reply]
Does it work better if you replace the sched_yield() call in src/unix/async. to a nanosleep() call with a 1 ns timeout? That should let it yield to lower prio threads.
```diff
diff --git a/src/unix/async.c b/src/unix/async.c
index e1805c32..cebbee09 100644
--- a/src/unix/async.c
+++ b/src/unix/async.c
@@ -32,7 +32,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
-#include <sched.h>  /* sched_yield() */
+#include <time.h>
 
 #ifdef __linux__
 #include <sys/eventfd.h>
@@ -107,7 +107,8 @@ static int uv__async_spin(uv_async_t* handle) {
      * inside the critical section and if it's running on the same CPU
      * as us, we'll just burn CPU cycles until the end of our time slice.
      */
-    sched_yield();
+    struct timespec t = {0, 1};
+    nanosleep(&t, NULL);
   }
 }
 
````

**Reply by xiaoxiang781216** [reply]
It's definitely work we have tried before, but I want to fix the problem from the root. 
Basically, it's impossible to fix "free and use" in libuv level: any change just reduce the race window, but the race still exist there. We should explicitly highlight the race in the document instead.

**Reply by bnoordhuis** [reply]
What do you suggest? Reverting #2231 is not on the table, that fixed a real bug that people were hitting.

**Reply by xiaoxiang781216** [reply]
You mean the bug described in https://github.com/libuv/libuv/issues/2226?

1. If you want to fix "use after free of uv_async_t" in the chain of event: uv_async_send->async callback->uv_close(or even uv_loop_delete),  I think the right fix is: https://github.com/libuv/libuv/pull/2656.

2. If you want to fix "use after free of uv_async_t" in general, it is impossible inside libuv. The caller has to coordinate the close and the asynchronous operation by self.

Actually, https://github.com/libuv/libuv/issues/2226 is the first case, which is the real race condition bug that libuv should fix.

**Reply by bnoordhuis** [reply]
(2) indeed isn't under discussion.

As to (1): I assume you've seen the discussion in #2656 and why it wasn't merged. Why then "right fix"?

**Reply by xiaoxiang781216** [reply]
Yes, you are right. #2656 can fix the race condition of handle close, but not looper deletion.
The solution I can imagine:

1. Fetch async_wfd before setting pending to 1, and let OS catch the stale fd for us
2. Call dup on async_wfd and close it after write return
3. Add refcount to uv_loop_t and hold the reference temporarily during sending the request

#2226 only concern about the close of uv_async_t in callback which is a common usage pattern. But, it's very rare that user will destroy uv_looper_t in callback. So, item 1 look a good balance between the performance and the behaviour.

What your opinion?

**Reply by bnoordhuis** [reply]
I get the impression you have the wrong idea about what #2231 fixes. It sounds like you think it's to accommodate user error: closing a handle that's used by another thread.

What it actually fixes is that `uv_async_send()` should logically be an atomic operation but wasn't, and that was observable by other threads.

The priority deadlock you mentioned is something we can investigate if it's an issue for you.

**Reply by xiaoxiang781216** [reply]
> I get the impression you have the wrong idea about what #2231 fixes. It sounds like you think it's to accommodate user error: closing a handle that's used by another thread.
> 

I just consider the user close handle or looper in the loop thread now, not a general thread after the discussion yesterday.

> What it actually fixes is that `uv_async_send()` should logically be an atomic operation but wasn't, and that was observable by other threads.
> 

If we don't consider the "use after free", I think that it isn't an issue that the observation of pending=>1 before writing async_wfd.
If the goal of uv_async_send is an atomic operation, I think futex or mutex is better choice than the adhoc spinlock implementation, because OS always has the better knowledge and more context to synchronize/schedule the thread than application library.

> The priority deadlock you mentioned is something we can investigate if it's an issue for you.

Yes, it's my initial motivation, but I also want to avoid the busy loop as much as possible.

**Reply by bnoordhuis** [reply]
> If the goal of uv_async_send is an atomic operation, I think futex or mutex is better choice than the adhoc spinlock 

There's probably no space in the `uv_async_t` struct to make that possible. What's more, using a mutex was tried a long time ago (pre-v1.x) and it was literally orders of magnitude slower.

I mean, you're welcome to give it a try but you should prepare yourself for disappointment. `sizeof(uv_async_t)` cannot change and performance should be in the same ballpark.
